### PR TITLE
otelzap: Use pool for encoder

### PR DESCRIPTION
**Pull request by khushijain21**
Part of https://github.com/open-telemetry/opentelemetry-go-contrib/issues/5191

Pre-work https://github.com/open-telemetry/opentelemetry-go-contrib/pull/5279

Results:

```
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/contrib/bridges/otelzap
cpu: 12th Gen Intel(R) Core(TM) i5-1245U
                           │ old_bench.txt │            new_bench.txt             │
                           │    sec/op     │    sec/op     vs base                │
CoreWrite/10_fields-12       1170.5n ± 28%   511.8n ± 44%  -56.28% (p=0.000 n=10)
CoreWrite/20_fields-12        2.665µ ± 24%   1.441µ ± 20%  -45.93% (p=0.000 n=10)
CoreWrite/Namespace-12        2.057µ ± 29%   1.211µ ±  5%  -41.13% (p=0.000 n=10)
CoreWrite/With10_fields-12    249.3n ± 22%   186.4n ± 21%  -25.23% (p=0.015 n=10)
CoreWrite/With20_fields-12    716.5n ± 30%   513.4n ± 40%        ~ (p=0.072 n=10)
CoreWrite/WithNamespace-12    24.39n ± 22%   14.16n ± 20%  -41.93% (p=0.000 n=10)
geomean                       550.9n         326.4n        -40.74%

                           │ old_bench.txt  │             new_bench.txt              │
                           │      B/op      │     B/op      vs base                  │
CoreWrite/10_fields-12         978.0 ± 0%       307.0 ± 0%  -68.61% (p=0.000 n=10)
CoreWrite/20_fields-12        2180.0 ± 0%       818.0 ± 0%  -62.48% (p=0.000 n=10)
CoreWrite/Namespace-12       1.719Ki ± 0%     1.002Ki ± 0%  -41.70% (p=0.000 n=10)
CoreWrite/With10_fields-12     208.0 ± 0%       208.0 ± 0%        ~ (p=1.000 n=10) ¹
CoreWrite/With20_fields-12     640.0 ± 0%       640.0 ± 0%        ~ (p=1.000 n=10) ¹
CoreWrite/WithNamespace-12     0.000 ± 0%       0.000 ± 0%        ~ (p=1.000 n=10) ¹
geomean                                   ²                 -36.01%                ²
¹ all samples are equal
² summaries must be >0 to compute geomean

                           │ old_bench.txt │            new_bench.txt             │
                           │   allocs/op   │ allocs/op   vs base                  │
CoreWrite/10_fields-12        13.00 ± 0%     10.00 ± 0%  -23.08% (p=0.000 n=10)
CoreWrite/20_fields-12        22.00 ± 0%     17.00 ± 0%  -22.73% (p=0.000 n=10)
CoreWrite/Namespace-12        16.00 ± 0%     13.00 ± 0%  -18.75% (p=0.000 n=10)
CoreWrite/With10_fields-12    1.000 ± 0%     1.000 ± 0%        ~ (p=1.000 n=10) ¹
CoreWrite/With20_fields-12    1.000 ± 0%     1.000 ± 0%        ~ (p=1.000 n=10) ¹
CoreWrite/WithNamespace-12    0.000 ± 0%     0.000 ± 0%        ~ (p=1.000 n=10) ¹
geomean                                  ²               -11.42%                ²
¹ all samples are equal
² summaries must be >0 to compute geomean
```

### CODE REVIEW

**Comment by pellared** [comment]
Please make sure to add benchmarks to the PRs which are meant to improve performance.
You can also consider creating separate PRs that would contain just benchmarks.

**Comment by pellared** [comment]
Please use https://pkg.go.dev/golang.org/x/perf/cmd/benchstat to show the performance improvement (between main and this branch) in the PR description.

File: bridges/otelzap/encoder.go [hunk 1]
```diff
@@ -196,15 +230,13 @@ type arrayEncoder struct {
 }
 
 func (a *arrayEncoder) AppendArray(v zapcore.ArrayMarshaler) error {
-	// TODO: Use arrayEncoder from a pool.
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply bridges/otelzap/encoder.go:199]: What benchmark was used to see this?

[the remark this answers is not in the document]
Comment (khushijain21) [reply bridges/otelzap/encoder.go:199]: not entirely sure why, but using pool in `AppendArray` and `AppendObject` seems to be causing shared memory issue. I visualised one of the failures with nested cases. For now, I avoid using pool for `AppendArray` and `AppendObject`. 

Capture3

**Comment by khushijain21** [comment]
@pellared

File: bridges/otelzap/encoder.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pellared) [reply bridges/otelzap/encoder.go:199]: At quick glance, I think that the issue is caused by the fact that in array encoder `free` function does:

```
arr.elems = arr.elems[:0]
```

which causes to reuse the same underlying array. The next `ArrayAppender` is going to access the same elements when doing

```
err := v.MarshalLogArray(arr)
```

before the logs are emitted

I suggest not using the pooled encoder for `AppendArray` and `AppendObject` in scope of this PR.

I think (hope) that `AppendArray` and `AppendObject` are not going to be frequently invoked/used by the bridge users.

File: bridges/otelzap/encoder.go [hunk 2]
```diff
@@ -6,13 +6,47 @@ package otelzap // import "go.opentelemetry.io/contrib/bridges/otelzap"
...
+	New: func() interface{} {
+		// Similar to console_encoder which uses capacity of 2
+		return &arrayEncoder{elems: make([]log.Value, 0, 2)}
+	},
+}
+
+func getArrayEncoder() (arr *arrayEncoder, free func()) {
+	arr = arrayEncoderPool.Get().(*arrayEncoder)
+	return arr, func() {
+		arr.elems = arr.elems[:0]
```
Comment (pellared) [inline bridges/otelzap/encoder.go:27]: Can you add this comment?

https://github.com/open-telemetry/opentelemetry-go-contrib/blob/c80e46445e3e1726dda9e9fc0c34e255f2ff908e/bridges/otelslog/handler.go#L352-L354

File: bridges/otelzap/encoder.go [hunk 3]
```diff
@@ -6,13 +6,47 @@ package otelzap // import "go.opentelemetry.io/contrib/bridges/otelzap"
...
+	New: func() interface{} {
+		// Similar to console_encoder which uses capacity of 2
+		return newObjectEncoder(2)
+	},
+}
+
+func getObjectEncoder() (obj *objectEncoder, free func()) {
+	obj = objectEncoderPool.Get().(*objectEncoder)
+	return obj, func() {
+		obj.root.attrs = obj.root.attrs[:0]
```
Comment (pellared) [inline bridges/otelzap/encoder.go:42]: Can you add this comment?

https://github.com/open-telemetry/opentelemetry-go-contrib/blob/c80e46445e3e1726dda9e9fc0c34e255f2ff908e/bridges/otelslog/handler.go#L352-L354

**Comment by pellared** [comment]
Can you please fix the build?

See: https://github.com/open-telemetry/opentelemetry-go-contrib/actions/runs/9854291298/job/27206720483?pr=5719

File: bridges/otelzap/encoder.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pellared) [reply bridges/otelzap/encoder.go:199]: I think the same problem with using shared memory applies for `AddArray` and `AddObject`. Similarly to https://github.com/open-telemetry/opentelemetry-go-contrib/issues/5879

**Pull request closed without merging**